File: thanh123456_AcceptedCodeVNSPOJ/BONUS13.pas
uses math;
const
     fin='';fon='';
     di:Array [1..8] of longint =(-2,-1,1,2,2,1,-1,-2);
     dj:array [1..8] of longint =(1,2,2,1,-1,-2,-2,-1);
var
     top1,top2,n:longint;
     a1,a2,b1,b2:array [0..1000] of longint;
     c:array [0..9,0..9] of int64;
     Hau,Xe,Voi,Ngua:array [0..64,0..64] of boolean;
procedure nhap;
     begin
          assign(input,fin);reset(input);
          assign(output,fon);rewrite(output);
          readln(n);
     end;
procedure taomang;
     var
          i,j,u,v:longint;
     begin
          for i:=1 to n do
               begin
                    readln(u,v,c[u,v]);
                    inc(top1);
                    a1[top1]:=u;
                    b1[top1]:=v;
               end;
          for i:=1 to 8 do
               for j:=1 to 8 do
                    begin
                         if c[i,j]=0 then
                              begin
                                   inc(top2);
                                   a2[top2]:=i;
                                   b2[top2]:=j;
                              end;
                    end;
     end;
Function Cheo1(i,j,u,v:longint):boolean;
     begin
          if i+j=u+v then exit(true)
          else exit(false);
     end;
Function Cheo2(i,j,u,v:longint):boolean;
     begin
          if i-j=u-v then exit(true)
          else exit(false);
     end;
Function Hang(i,j,u,v:longint):boolean;
     begin
          if i=u then exit(true)
          else exit(false);
     end;
Function Cot(i,j,u,v:longint):boolean;
     begin
          if j=v then exit(true)
          else exit(false);
     end;
Function Nhay(i,j,u,v:longint):boolean;
     var
          k:longint;
     begin
          for k:=1 to 8 do
               if (i+di[k]=u) and (j+dj[k]=v) then exit(true);
          exit(false);
     end;
procedure Init;
     var
          i,j:longint;
     begin
          for i:=1 to top1 do
               for j:=1 to top2 do
                    begin
                         if (Cheo1(a1[i],b1[i],a2[j],b2[j])=true) or (Cheo2(a1[i],b1[i],a2[j],b2[j])=true) then
                              begin
                                   Hau[i,j]:=true;
                                   Voi[i,j]:=true;
                              end;
                         if (Hang(a1[i],b1[i],a2[j],b2[j])=true) or (Cot(a1[i],b1[i],a2[j],b2[j])=true) then
                              begin
                                   hau[i,j]:=true;
                                   Xe[i,j]:=true;
                              end;
                         if Nhay(a1[i],b1[i],a2[j],b2[j])=true then
                              Ngua[i,j]:=true;
                    end;
     end;
procedure solve;
     var
          i,j,u,v,k:longint;
          res,s:int64;
     begin
          res:=0;
          for i:=1 to top2 do
               for j:=1 to top2 do
                    begin
                         if i=j then continue;
                         for u:=1 to top2 do
                              begin
                                   if (u=i) or (u=j) then continue;
                                   for v:=1 to top2 do
                                        begin
                                             if (v=i) or (v=j) or (v=u) then continue;
                                             s:=0;
                                             for k:=1 to top1 do
                                                  begin
                                                       if (Hau[k,i]=true) or (Voi[k,j]=true) or (Xe[k,u]=true) or (Ngua[k,v]=true) then
                                                            s:=s+c[a1[k],b1[k]];
                                                  end;
                                             res:=max(res,s);
                                        end;
                              end;
                    end;
          writeln(res);
     close(input); close(output);
     end;
begin
     Nhap;
     TaoMang;
     Init;
     solve;
end.

File: thanh123456_AcceptedCodeVNSPOJ/BONUS13.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1e2 + 10;
const int di[] = {-2,-1,1,2,2,1,-1,-2};
const int dj[] = {1,2,2,1,-1,-2,-2,-1};
int n,top1 = 0,top2 = 0;
struct ds
{
    int x,y;
};
ds a[N],b[N];
int c[N][N];
bool Hau[N][N],Tuong[N][N],Xe[N][N],Ma[N][N];
void nhap()
{
    scanf("%d",&n);
    for (int i = 1; i <= n; i++) {
        int x,y,d;
        scanf("%d %d %d",&x,&y,&d);
        c[x][y] = max(c[x][y],d);
    }
    for (int i = 1; i <= 8; i++) {
        for (int j = 1; j <= 8; j++) {
            if (c[i][j] == 0) {
                top1++;
                a[top1].x = i;
                a[top1].y = j;
            }
            else {
                top2++;
                b[top2].x = i;
                b[top2].y = j;
            }
        }
    }
}
bool Cheo(int x1,int y1,int x2,int y2)
{
    if (y1 - x1 == y2 - x2) {return true;}
    if (x1 + y1 == x2 + y2) {return true;}
    return false;
}
bool Ngang(int x1,int y1,int x2,int y2)
{
    if (x1 == x2) {return true;}
    else {return false;}
}
bool Cot(int x1,int y1,int x2,int y2)
{
    if (y1 == y2) {return true;}
    else {return false;}
}
bool ma(int x1,int y1,int x2,int y2)
{
    for (int i = 0; i < 8; i++) {
        if (x1 + di[i] == x2 && y1 + dj[i] == y2) {return true;}
    }
    return false;
}
void init()
{
    for (int i = 1; i <= top1; i++) {
        for (int j = 1; j <= top2; j++) {
            int u = a[i].x;
            int v = a[i].y;
            int x = b[j].x;
            int y = b[j].y;
            if (Cheo(u,v,x,y)) {
                Hau[j][i] = true;
                Tuong[j][i] = true;
            }
            if (Cot(u,v,x,y) || Ngang(u,v,x,y)) {
                Hau[j][i] = true;
                Xe[j][i] = true;
            }
            if (ma(u,v,x,y)) {
                Ma[j][i] = true;
            }
        }
    }
}
void solve()
{
    long long res = 0;
    res = 0;
    for (int i = 1; i <= top1; i++) {
        for (int j = 1; j <= top1; j++) {
            if (i == j) {continue;}
            for (int u = 1; u <= top1; u++) {
                if (u == i || u == j) {continue;}
                for (int v = 1; v <= top1; v++) {
                    if (v == i || v == j || v == u) {continue;}
                    long long gt = 0;
                    for (int k = 1; k <= top2; k++) {
                        if (Hau[k][i] || Tuong[k][j] || Xe[k][u] || Ma[k][v]) {
                            gt += 1ll * c[b[k].x][b[k].y];
                        }
                    }
                    if (gt > res) {res = gt;}
                }
            }
        }
    }
    cout<<res;
}
int main()
{
    nhap();
    init();
    solve();
}
